src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,295 @@
+            this.aggregatorDictionary = new ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator>();
+
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            this.snapshotTask = new TaskTimer() { Delay = GetWaitTime() };
+            this.snapshotTask.Start(this.SnapshotAndReschedule);
+        }
+
+        /// <summary>
+        /// Gets a list of metric processors associated
+        /// with this instance of <see cref="MetricAggregatorManager"/>.
+        /// </summary>
+        internal IList<IMetricProcessor> MetricProcessors
+        {
+            get
+            {
+                TelemetryConfiguration config = this.telemetryClient.TelemetryConfiguration;
+
+                return config.MetricProcessors;
+            }
+        }
+
+        /// <summary>
+        /// Creates single time series metric data aggregator.
+        /// </summary>
+        /// <param name="metricName">Name of the metric.</param>
+        /// <param name="dimensions">Optional dimensions.</param>
+        /// <returns>Value aggregator for the metric specified.</returns>
+        public MetricAggregator CreateMetricAggregator(string metricName, IDictionary<string, string> dimensions = null)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            return new MetricAggregator(this, metricName, dimensions);
+        }
+
+        /// <summary>
+        /// Flushes the in-memory aggregation buffers.
+        /// </summary>
+        public void Flush()
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            this.Flush(false);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            this.disposed = true;
+
+            if (this.snapshotTask != null)
+            {
+                this.snapshotTask.Dispose();
+            }
+
+            this.Flush(true);
+        }
+
+        internal SimpleMetricStatisticsAggregator GetStatisticsAggregator(MetricAggregator aggregator)
+        {
+            if (this.disposed)
+            {
+                throw new ObjectDisposedException(nameof(MetricAggregatorManager));
+            }
+
+            if (aggregator == null)
+            {
+                return null;
+            }
+
+            return this.aggregatorDictionary.GetOrAdd(aggregator, (aid) => { return new SimpleMetricStatisticsAggregator(); });
+        }
+
+        /// <summary>
+        /// Calculates wait time until next snapshot of the aggregators.
+        /// </summary>
+        /// <returns>Wait time.</returns>
+        private static TimeSpan GetWaitTime()
+        {
+            DateTimeOffset currentTime = DateTimeOffset.UtcNow;
```

Comment:
How are we testing the `GetWaitTime` logic? We have the `IClock` interface that can be injected instead of taking a direct dependency on the static `DateTimeOffset.UtcNow`.